amads/time/meter/__init__.py

```diff
@@ -0,0 +1,619 @@
+
+    Likewise, the minimum_pulse must be set to one of these values
+    (default = 64 for 64th note) and not be longer than the meter.
+
+    Finally, although we support and provide defaults for time signatures in the form "2+3/8",
+    there is no such support for more than one "/" (i.e., the user must build cases like "4/4 + 3/8" explicitly).
+    """
+
+    # Prep and checks
+    numerator, denominator = ts_str.split("/")
+    numerators = [
+        int(x) for x in numerator.split("+")
+    ]  # Note: support "+" before one "/", e.g., `2+3/4`
+    denominator = int(denominator)
+
+    supported_denominators = [1, 2, 4, 8, 16, 32, 64]
+    if denominator not in supported_denominators:
+        raise ValueError(
+            "Invalid time signature denominator; chose from one of:"
+            f" {supported_denominators}."
+        )
+    if minimum_pulse not in supported_denominators:
+        raise ValueError(
+            "Invalid minimum_pulse: it should be expressed as a note value, from one of"
+            f" {supported_denominators}."
+        )
+
+    # Prep. "hiddenLayer"/s
+    hidden_layer_mappings = (
+        ([4], [2, 2]),
+        ([6], [3, 3]),
+        ([9], [3, 3, 3]),  # alternative groupings need to be set out, e.g., 2+2+2+3
+        ([12], [[6, 6], [3, 3, 3, 3]]),  # " e.g., 2+2+2+3
+        ([15], [3, 3, 3, 3, 3]),  # " e.g., 2+2+2+3
+        ([6, 9], [[6, 9], [3, 3, 3, 3, 3]]),
+        ([9, 6], [[6, 6], [3, 3, 3, 3, 3]]),
+    )
+
+    if enforce_2s_3s:
+        for h in hidden_layer_mappings:
+            if numerators == h[0]:
+                numerators = h[1]
+
+    if isinstance(numerators[0], list):
+        sum_num = sum(numerators[0])
+    else:
+        sum_num = sum(numerators)
+
+    measure_length = sum_num * 4 / denominator
+
+    # Now ready for each layer in order:
+
+    # 1. top level = whole cycle (always included as entry[0])
+    start_hierarchy = [[0.0, measure_length]]
+
+    # 2. "hidden" layer/s
+    if len(numerators) > 1:  # whole cycle covered.
+        if isinstance(numerators[0], list):
+            for level in numerators:
+                starts = starts_from_beat_pattern(level, denominator)
+                start_hierarchy.append(starts)
+        else:
+            starts = starts_from_beat_pattern(numerators, denominator)
+            start_hierarchy.append(starts)
+
+    # 3. Finally, everything at the denominator level and shorter:
+    this_denominator_power = supported_denominators.index(denominator)
+    max_denominator_power = supported_denominators.index(minimum_pulse)
+
+    for this_denominator in supported_denominators[
+        this_denominator_power : max_denominator_power + 1
+    ]:
+        this_ql = 4 / this_denominator
+        starts = starts_from_lengths(measure_length, this_ql)
+        start_hierarchy.append(starts)
+
+    return start_hierarchy
+
+
+def starts_from_time_signature_and_levels(
+    ts_str: str,
+    levels: Union[list, None] = None,
```

Comment:
Are you sure you want `levels` to be a list of all levels to be included? so that users can skip consecutive levels? might it be more natural to rename this `depth` and make this be the total number of levels to include?

Reply:
I agree that depth is more common, but I prefer the more flexible option in this case.

While the combination of  `4` and `1` tends to be taken to imply an intervening `2`, I see a value in being able to code and work with the former (without the 'fill in') as a metrical structure in itself. This connects with general considerations like the usage of levels in practice, and specific cases like the distinction between 4/4 and 2/2 and the presence or absence of a half-bar division there.